cassiopeia/data/CassiopeiaTree.py

```diff
@@ -0,0 +1,357 @@
+        structure: pruning lineages, collapsing mutationless edges, and
+        collapsing unifurcations
+    TODO(mattjones315): Add utility methods to compute the colless index
+        and the cophenetic correlation wrt to some cell meta item
+
+    Args:
+        character_matrix: The character matrix for the lineage.
+        cell_meta: Per-cell meta data
+        character_meta: Per-character meta data
+        priors: A dictionary storing the probability of a character mutating
+            to a particular state.
+        tree: A tree for the lineage. 
+    """
+
+    def __init__(
+        self,
+        character_matrix: pd.DataFrame,
+        cell_meta: Optional[pd.DataFrame] = None,
+        character_meta: Optional[pd.DataFrame] = None,
+        priors: Optional[Dict[int, Dict[int, float]]] = None,
+        tree: Optional[Union[str, ete3.Tree, nx.DiGraph]] = None
+    ):
+
+        self.character_matrix = character_matrix
+        self.cell_meta = cell_meta
+        self.character_meta = character_meta
+        self.priors = priors
+        self.__network = None
+
+        self.populate_tree(tree)
+
+    def populate_tree(self, tree: Union[str, ete3.Tree, nx.DiGraph]):
+
+        if isinstance(tree, nx.DiGraph):
+            self.__network = tree
+        if isinstance(tree, str):
+            self.__network = utilities.newick_to_networkx(tree)
+        if isinstance(tree, ete3.Tree):
+            self.__network = utilities.ete3_to_networkx(tree)
+
+        if self.__network is not None:
+
+            # add character states
+            for n in self.nodes:
+                if n in self.character_matrix.index.values:
+                    self.__network.nodes[n]['character_states'] = self.character_matrix.loc[n].to_list()
+                else:
+                    self.__network.nodes[n]['character_states'] = []
+            
+            # instantiate branch lengths
+            for u,v in self.edges:
+                self.__network[u][v]['length'] = 1
+
+            # instantiate node ages and edge depth
+            self.__network.nodes[self.root]['age'] = 0
+            self.__network.nodes[self.root]['edge_depth'] = 0
+            for u,v in self.depth_first_traverse_edges(source=self.root):
+                self.__network.nodes[v]['age'] = self.__network.nodes[u]['age'] + self.__network[u][v]['length']
+                self.__network.nodes[v]['edge_depth'] = self.__network.nodes[v]['age']
+
+    @property
+    def n_cell(self) -> int:
+        """Returns number of cells in character matrix.
+        """
+        return self.character_matrix.shape[0]
+
+    @property
+    def n_character(self) -> int:
+        """Returns number of characters in character matrix.
+        """
+        return self.character_matrix.shape[1]
+
+    @property
+    def root(self) -> str:
+        """Returns root of tree.
+        """
+        if self.__network is None:
+            return None
```

Comment:
I had this same design concern when I was developing. I was pretty up in the air on the design, so because we both had the same thought I'm going to go ahead and change the behavior here. Thanks for the suggestion!